app/src/main/kotlin/compiler/intermediate_form/Instruction.kt

```diff
@@ -1,61 +1,152 @@
+        data class SubRR(override val reg0: Register, override val reg1: Register) : TwoRegIns() // SUB  reg0, reg1   reg0 <- reg0 - reg1
+        data class MulRR(override val reg0: Register, override val reg1: Register) : TwoRegIns() // IMUL reg0, reg1   reg0 <- reg0 * reg1
+        data class DivR(val reg: Register) : InPlaceInstruction() { //                              IDIV reg          RAX  <- RDX:RAX / reg
+            override val regsDefined: Collection<Register> = setOf(Register.RAX, Register.RDX) //                     RDX  <- RDX:RAX % reg
+            override val regsUsed: Collection<Register> = setOf(Register.RAX, Register.RDX, reg)
+        }
+
+        // Bitwise instructions
+        data class NotR(override val reg: Register) : OneRegIns() //                                NOT  reg          reg  <- ~reg
+        data class AndRR(override val reg0: Register, override val reg1: Register) : TwoRegIns() // AND  reg0, reg1   reg0 <- reg0 & reg1
+        data class OrRR(override val reg0: Register, override val reg1: Register) : TwoRegIns() //  OR   reg0, reg1   reg0 <- reg0 & reg1
+        data class XorRR(override val reg0: Register, override val reg1: Register) : TwoRegIns() // XOR  reg0, reg1   reg0 <- reg0 ^ reg1
+        data class ShiftLeftR(val reg: Register) : InPlaceInstruction() { //                        SAL  reg,  CL     reg  <- reg << CL
+            override val regsDefined: Collection<Register> = setOf(reg)
+            override val regsUsed: Collection<Register> = setOf(Register.RCX)
+        }
+        data class ShiftRightR(val reg: Register) : InPlaceInstruction() { //                       SAR  reg,  CL     reg  <- reg >> CL
+            override val regsDefined: Collection<Register> = setOf(reg)
+            override val regsUsed: Collection<Register> = setOf(Register.RCX)
+        }
+
+        // Comparison instructions
+        data class CmpRR(val reg0: Register, val reg1: Register) : InPlaceInstruction() { //  CMP   reg0, reg1
+            override val regsUsed: Collection<Register> = setOf(reg0, reg1)
+        }
+        data class TestRR(val reg0: Register, val reg1: Register) : InPlaceInstruction() { // TEST  reg0, reg1
+            override val regsUsed: Collection<Register> = setOf(reg0, reg1)
+        }
+
+        // SETcc instructions
+        sealed class SetInstruction : InPlaceInstruction() {
+            abstract val reg: Register
+            override val regsDefined: Collection<Register> get() = setOf(reg)
+            // I'm unsure if this is needed
+            // The SETcc instructions set the lower 8 bits of a register
+            // so the value after SETcc is dependent on the value of reg before instruction.
+            // Specifically, if this wasn't included in when doing
+            // 1: XOR    reg,   reg   ; reset to zero
+            // 2: CMP    rega,  regb
+            // 3: SETcc  reg          ; set lower 8 bits depending on comparison
```

Comment:
Docs say (and GCC confirms that, so it'll probably be the same with NASM) that operation SETcc will not accept 64-bit register at all.

Reply:
Yeah, I know that SETcc takes 8-bit registers. I meant for Setcc(reg) to represent SETcc with low 8-bit register corresponding to reg as an operand. I read in documentation that the upper bits are unchanged for 8-bit operands. I'm not sure if the byte order is correct (so setting AL to 1 is the same as setting RAX to 1). I haven't tested it and I'm not all that versed in assembly.

Reply:
According to AMD64 docs, smaller registers (such as AL, AX, EAX) occupy lower bits of full register, so the byte order should be correct.